### ISSUE TITLE
`CategoricalAccuracy` metric crashes on GPU

### ISSUE THREAD

**Issue by joelgrus** [allenai/allennlp#166]
the culprit seems to be 

https://github.com/allenai/allennlp/blob/master/allennlp/training/metrics/categorical_accuracy.py#L64

because `mask` is on the GPU but `count` is on the CPU.

I can think of a few ways to fix this, but I'm not sure which is least hacky, nor if this problem exists elsewhere

**Comment by matt-gardner** [comment]
What's the context?  I use `CategoricalAccuracy` in BiDAF, and it works fine.

**Comment by joelgrus** [comment]
hmm, it's the `SimpleTagger` model (the one in the tutorial). it crashed when I tried to train it on a GPU machine. when I changed the code to

```
count = torch.ones(gold_labels.size()).cuda()
```

it worked

**Comment by matt-gardner** [comment]
Ah, it works in BiDAF because I don't pass a mask, so I don't hit that line.  Yeah, that does seem like it should break.  The correct way to fix this is here: https://github.com/allenai/allennlp/blob/master/allennlp/nn/util.py#L58-L61.

**Comment by joelgrus** [comment]
is there a reason that's better than

```
count = torch.ones(gold_labels.size())
if torch.cuda.available():
    count = count.cuda()
```

based on https://github.com/pytorch/pytorch/issues/1668

**Comment by matt-gardner** [comment]
Yes.  That code will break if you have a GPU available, but you're running on the CPU.
